share/wake/lib/core/json.wake

```diff
@@ -538,3 +538,243 @@ export def lastValueInList (values: List JValue): Result JValue Error =
+    x         = x
+
+# If the input is a number, returns a singelton list
+# containing that number as a double. This can be used
+# to extract a number from a JValue or in list operation
+# chains as a safe JValue to number conversion.
+#
+# This works with both JDouble and JInteger. An integer
+# will be converted to a double if possible. If the
+# conversion is not possible because the integer is
+# out of range jNumber will return just Nil
+#
+# Examples:
+# ```
+#   jNumber (JDouble 3.14) = 3.14, Nil
+#   jNumber (JInteger 10) = 10.0, Nil
+#   jNumber (JString "Mal") = Nil
+#   jNumber (JInteger (factorial 100)) = Nil
+#   single json |> jField "height" |> jNumber = 1.75, Nil
+#   single json |> jField "age" |> jNumber = 28.0, Nil
+#   single json |> jField "name" |> jNumber = Nil
+# ```
+export def jNumber: JValue => List Double = match _
+    JDouble x  = single x
+    JInteger x = single x | mapPartial dint
+    _          = Nil
+
+# If `input` is a number, updates the number using `editFn`.
+# This will edit both JDouble and JInteger. The JInteger will
+# be converted to a JDouble if the conversion from Integer
+# to Double succeeds. If the conversion from Integer to
+# double does not succeed the integer remains unedited.
+#
+# Parameters:
+#   - `editFn`: The function to update a number with
+#   - `input`: The JValue to be modified
+#
+# Examples:
+# ```
+#   editJNumber (_+.1.0) (JDouble 10.0) = JDouble 11.0
+#   editJNumber (_+.1.0) (JInteger 10) = JDouble 11.0
+#   editJNumber (_+.1.0) (JString "Bob") = JString "Bob"
+#   editJNumber (_+.1.0) (JInteger (factorial 100)) = JInteger (factorial 100)
+#   def newTallest: (stats: JValue) => JValue = editJField "tallestBuilding" $ editJNumber (\_ 1000.0)
+#   def grow: (person: JValue) => JValue = editJField "height" $ editJNumber (_+.0.02)
+# ```
+export def editJNumber (f: Double => Double): JValue => JValue = match _
+    JDouble x  = JDouble (f x)
+    JInteger x = match (dint x)
+        Some dx = JDouble (f dx)
+        None    = JInteger x
+    x          = x
+
+
+export def jNull: JValue => List Unit = match _
+    JNull = single Unit
+    _     = Nil
+
+export def editJNull (f: Unit => Unit): JValue => JValue = match _
+    JNull =
+        def _ = f Unit # f could have an effect so we can't eliminate it
+        JNull
+    x     = x
+
+export def jObject: JValue => List (List (Pair String JValue)) = match _
+    JObject obj = single obj
+    _           = Nil
+
+export def editJObject (f: List (Pair String JValue) => List (Pair String JValue)): JValue => JValue = match _
+    JObject obj = JObject (f obj)
+    x           = x
+
+export def jArray: JValue => List (List JValue) = match _
+    JArray arr = single arr
+    _          = Nil
+
+export def editJArray (f: List JValue => List JValue): JValue => JValue = match _
+    JArray arr = JArray (f arr)
+    x          = x
+
+export def jFlatten: JValue => List JValue = match _
+  JArray a = a
+  _ = Nil
+
```

Comment:
We are missing the `/../` operator from the old API --> the give me every node of the subtree operation.
So you can find all nodes in the tree with some property.

Reply:
Easy to add!

In Haskell land this operation is called "plate" but that doesn't mean anything outside of those places: https://hackage.haskell.org/package/lens-5.1/docs/Control-Lens-Plated.html, https://www.cs.york.ac.uk/fp/darcs/uniplate/uniplate.htm but "jPlate"would be the best name for me personally, I just don't think that's an empathetic name for our users.

jChildren maybe? jSub sounds like subtraction? jNodes sounds too generic to me. I think I'm leaning towards jChildren.

Reply:
Perhaps `jDescendants`?

Reply:
jDescendants implies all children recursively to me rather than just all immediate children. Like "I'm a descendant of someone who speaks German" is a true statement but "I'm a child of someone who speaks German" is not a true statement (just to illustrate the difference)

Reply:
Perhaps I wasn't clear. All children recursively is what I'm asking for. That's what `/../` did.
So, I think now I am onboard with TWO functions: `jChildren` (for JObject=>values) and `jDescendants`.

Reply:
(Though, I'm not sure `jChildren` is actually useful)

Reply:
Ok cool. Lets just add jDescendants for now then. I agree that's a good name for that since I managed to figure out the semantics of it despite assuming you meant something else :)

For this change lets just get the minimum we know is useful, we can add new helper functions in future PRs or wake versions. After we land this change we should probably consider additional missing features and see if we can spot any patterns, maybe start to rewrite some code in Federation with the new library and see what emergent patterns we see.